### PR TITLE
Cce gh executable interpolator utilities

### PR DESCRIPTION
**Pull request by moxcodes**
## Proposed changes

The interpolation changes and additions necessary for the combined GH-CCE executable.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivClean.hpp [hunk 1]
```diff
@@ -131,30 +132,7 @@ class CProxy_ConstGlobalCache;
...
-      hydro::Tags::MassFluxCompute<DataVector, 3, ::Frame::Inertial>>;
-  using compute_target_points =
-      intrp::TargetPoints::KerrHorizon<KerrHorizon, ::Frame::Inertial>;
-  using post_interpolation_callback =
-      intrp::callbacks::ObserveTimeSeriesOnSurface<tags_to_observe, KerrHorizon,
-                                                   KerrHorizon>;
-};
-
-template <typename InitialData, typename...InterpolationTargetTags>
+template <typename InitialData, typename... InterpolationTargetTags>
```
Comment (geoffrey4444) [inline src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivClean.hpp:134]: Why did you move this struct? Just wondering

**Review by geoffrey4444 — CHANGES_REQUESTED** [review]

File: src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivClean.hpp (hunk 1 above)
Comment (moxcodes) [reply src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivClean.hpp:134]: The original version of getting the interpolation points to the elements required passing around the pdal of the dg element array, which I thought was inelegent, so I've added the type alias `interpolating_dg_component` to the interpolation tags that would need access to that type. However, that now requires the dg element array component type to exist before I define the interpolation tag struct, so it forced the struct definition to follow the metavariables definition.

File: src/NumericalAlgorithms/Interpolation/Actions/InterpolateToTarget.hpp [hunk 2]
```diff
@@ -53,6 +60,11 @@ struct InterpolateToTarget {
       Parallel::ConstGlobalCache<Metavariables>& cache,
       const ArrayIndex& array_index, const ActionList /*meta*/,
       const ParallelComponent* const /*meta*/) noexcept {
+    // Exit early if we are on a substep and we don't want to interpolate
+    // substeps
+    if (not InterpolationTargetTag::should_interpolate(box)) {
```
Comment (markscheel) [inline src/NumericalAlgorithms/Interpolation/Actions/InterpolateToTarget.hpp]: Is every InterpolationTargetTag now required to have a should_interpolate function?  (i.e. if it is missing one gets a compiler error?)  Would it be better to make it optional?  Also, since this changes the interface to InterpolationTarget, it should be documented near the top of InterpolationTarget.hpp.

Also, the comment mentions substeps but the name 'should_interpolate' seems to be completely general and has nothing to do with substeps.  Not sure if the comment should be changed or if the function should be renamed. Maybe replace the name 'should_interpolate' with 'should_interpolate_at_this_temporal_id' or 'should_interpolate_on_this_call' or something like that?

File: src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivClean.hpp [hunk 3]
```diff
@@ -398,6 +376,32 @@ struct EvolutionMetavars {
...
+  using compute_items_on_target = tmpl::push_front<
+      tags_to_observe,
+      StrahlkorperTags::EuclideanAreaElement<::Frame::Inertial>,
+      hydro::Tags::MassFluxCompute<DataVector, 3, ::Frame::Inertial>>;
+  using compute_target_points =
+      intrp::TargetPoints::KerrHorizon<KerrHorizon, ::Frame::Inertial>;
+  using post_interpolation_callback =
+      intrp::callbacks::ObserveTimeSeriesOnSurface<tags_to_observe, KerrHorizon,
+                                                   KerrHorizon>;
+  using interpolating_dg_component =
```
Comment (markscheel) [inline src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivClean.hpp]: Is InterpolationTargetTag now required to have this type alias?  Should be documented in InterpolationTarget.hpp near the top (including under what conditions it is required).

File: tests/Unit/NumericalAlgorithms/Interpolation/Test_InterpolateToTarget.cpp [hunk 4]
```diff
@@ -78,13 +78,21 @@ struct initialize_elements_and_queue_simple_actions {
...
     }
   }
 };
 
-template <bool HaveComputeItemsOnSource, bool AddComputeItemToBox>
+template <bool HaveComputeItemsOnSource, bool AddComputeItemToBox,
+          bool ShouldInterpolate>
 struct MockMetavariables {
   static constexpr bool add_compute_item_to_box = AddComputeItemToBox;
+  static constexpr bool should_interpolate_setting = ShouldInterpolate;
```
Comment (markscheel) [inline tests/Unit/NumericalAlgorithms/Interpolation/Test_InterpolateToTarget.cpp]: Maybe rename 'should_interpolate_value' ?

**Review by markscheel — CHANGES_REQUESTED** [review]
just a few documentation things and possible renaming.  I like the removal of the pdal template argument in InterpolationTargetSendPoints; I felt dirty when I added it in the first place.

File: src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivClean.hpp (hunk 3 above)
Comment (moxcodes) [reply src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivClean.hpp]: this is only required if the interpolation target has `is_sequential` aliased to `std::false_type`. I'll add that documentation.

File: src/NumericalAlgorithms/Interpolation/Actions/InterpolateToTarget.hpp (hunk 2 above)
Comment (moxcodes) [reply src/NumericalAlgorithms/Interpolation/Actions/InterpolateToTarget.hpp]: Only the interpolation target tags that are used in this action need to have the type alias, and I think this action can only be used for the interpolation target tags that have `is_sequential` be false. I'll also make the function optional with a type traits. I'll update the InterpolationTarget documentation for this as well.

I think it's best to change the comment -- the idea is to have a very generic (trigger-like) boolean function to determine whether it's time to interpolate.

**Comment by moxcodes** [comment]
fixups posted and rebased on develop.
Cheers!

**Comment by moxcodes** [comment]
Rebased on develop, conflicts fixed.
Cheers!

File: src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivClean.hpp [hunk 5]
```diff
@@ -5,6 +5,7 @@
 
 #include <vector>
 
+#include "NumericalAlgorithms/Interpolation/Actions/ElementInitInterpPoints.hpp"
```
Comment (kidder) [inline src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivClean.hpp]: alphabet

File: src/NumericalAlgorithms/Interpolation/Actions/InterpolationTargetSendPoints.hpp [hunk 6]
```diff
@@ -47,7 +47,7 @@ struct InterpolationTargetSendTimeIndepPointsToElements {
     auto coords = InterpolationTarget_detail::block_logical_coords<
         InterpolationTargetTag>(box, tmpl::type_<Metavariables>{});
     auto& receiver_proxy = Parallel::get_parallel_component<
-        DgElementArray<Metavariables, PhaseDepActionList>>(cache);
+        typename InterpolationTargetTag::interpolating_dg_component>(cache);
```
Comment (kidder) [inline src/NumericalAlgorithms/Interpolation/Actions/InterpolationTargetSendPoints.hpp]: if `interpolating_dg_component` does not depend upon DG, then rename to `interpolating_component`

File: src/NumericalAlgorithms/Interpolation/InterpolationTargetKerrHorizon.hpp [hunk 7]
```diff
@@ -172,7 +182,25 @@ struct KerrHorizon {
     // the code that transforms coordinates from the Strahlkorper Frame
     // to Frame::Inertial will go here.  That transformation
     // may depend on `temporal_id`.
-    return db::get<StrahlkorperTags::CartesianCoords<::Frame::Inertial>>(box);
+    const auto& options =
```
Comment (kidder) [inline src/NumericalAlgorithms/Interpolation/InterpolationTargetKerrHorizon.hpp]: options -> kerr_horizon

File: src/NumericalAlgorithms/Interpolation/InterpolationTargetKerrHorizon.hpp [hunk 8]
```diff
@@ -83,6 +92,7 @@ struct KerrHorizon {
   std::array<double, 3> center{};
   double mass{};
   std::array<double, 3> dimensionless_spin{};
+  bool theta_varies_fastest{};
```
Comment (kidder) [inline src/NumericalAlgorithms/Interpolation/InterpolationTargetKerrHorizon.hpp]: theta_varies_fastest_memory_layout

File: src/NumericalAlgorithms/Interpolation/InterpolationTargetKerrHorizon.cpp [hunk 9]
```diff
@@ -14,11 +14,13 @@ namespace OptionHolders {
 KerrHorizon::KerrHorizon(size_t l_max_in, std::array<double, 3> center_in,
                          double mass_in,
                          std::array<double, 3> dimensionless_spin_in,
+                         const bool theta_varies_fastest_in,
```
Comment (kidder) [inline src/NumericalAlgorithms/Interpolation/InterpolationTargetKerrHorizon.cpp:15]: why is this needed?

File: src/NumericalAlgorithms/Interpolation/InterpolationTargetKerrHorizon.cpp [hunk 10]
```diff
@@ -19,6 +18,7 @@ KerrHorizon::KerrHorizon(size_t l_max_in, std::array<double, 3> center_in,
     : l_max(l_max_in),
       center(std::move(center_in)),  // NOLINT
       mass(mass_in),
+      // NOLINTNEXTLINE(performance-move-const-arg)
```
Comment (kidder) [inline src/NumericalAlgorithms/Interpolation/InterpolationTargetKerrHorizon.cpp]: don't move anything in this constructor and remove the NOLINTs

File: src/NumericalAlgorithms/Interpolation/InterpolationTargetDetail.hpp [hunk 11]
```diff
@@ -480,5 +480,16 @@ void set_up_interpolation(
       });
 }
 
+template <typename T, typename = std::void_t<>>
```
Comment (kidder) [inline src/NumericalAlgorithms/Interpolation/InterpolationTargetDetail.hpp]: just use macro in `Utilities/TypeTraits/CreateIsCallable.hpp`

File: src/NumericalAlgorithms/Interpolation/InterpolationTarget.hpp [hunk 12]
```diff
@@ -96,14 +96,19 @@ namespace intrp {
 ///      Domain will be filled with this value. If this variable is not defined,
 ///      then the `apply` function must check for invalid points,
 ///      and should typically exit with an error message if it finds any.
-/// - should_interpolate (optional):
+/// - should_interpolate (optional) :
```
Comment (kidder) [inline src/NumericalAlgorithms/Interpolation/InterpolationTarget.hpp]: why the new space?

**Review by kidder — CHANGES_REQUESTED** [review]
reviewed with @nilsdeppe

File: src/NumericalAlgorithms/Interpolation/InterpolationTargetKerrHorizon.cpp (hunk 9 above)
Comment (moxcodes) [reply src/NumericalAlgorithms/Interpolation/InterpolationTargetKerrHorizon.cpp:15]: It needs to be specifiable by options, because the ordering of points is different for the apparent horizon interface and for CCE -- or have I misunderstood the question?

Comment (nilsdeppe) [reply src/NumericalAlgorithms/Interpolation/InterpolationTargetKerrHorizon.cpp:15]: Nope, that was the confusion. Maybe document that just as a "for example" usage case so it's recorded somewhere?

**Comment by moxcodes** [comment]
fixups posted, rebased on develop.
Cheers!

File: src/NumericalAlgorithms/Interpolation/InterpolationTarget.hpp [hunk 13]
```diff
@@ -103,7 +103,12 @@ namespace intrp {
 ///      Often, this will examine the `temporal_id` to determine whether to
 ///      interpolate, but it may examine any part of the \ref DataBoxGroup.
 ///      If omitted, the behavior is as if the `should_interpolate` function
-///      always returns true.
+///      always returns true. This function is called using the DG element's
+///      \ref DataBoxGroup and is only used if `InterpolateToTarget` is
+///      in the element's action list.
+/// - interpolating_dg_component (required if not `is_sequential`):
```
Comment (kidder) [inline src/NumericalAlgorithms/Interpolation/InterpolationTarget.hpp]: rename to interpolating_component

**Review by kidder — COMMENTED** [review]
you can squash after @markscheel gives the go-ahead

File: src/NumericalAlgorithms/Interpolation/Actions/InterpolateToTarget.hpp [hunk 14]
```diff
@@ -30,13 +31,20 @@ namespace Actions {
...
 /// Uses:
 /// - DataBox:
 ///   - `intrp::Tags::InterpPointInfo<Metavariables>`
 ///   - `Tags::Mesh<Metavariables::volume_dim>`
 ///   - Variables tagged by
 ///     InterpolationTargetTag::vars_to_interpolate_to_target
+///   - `::Tags::TimeStepId`
+///   - `::Tags::HistoryEvolvedVariables<>`
+/// - ConstGlobalCache
+///   - `::Tags::TimeStepper<>`
```
Comment (markscheel) [inline src/NumericalAlgorithms/Interpolation/Actions/InterpolateToTarget.hpp]: Where are TimeStepper tags now used?

File: src/NumericalAlgorithms/Interpolation/InterpolationTarget.hpp [hunk 15]
```diff
@@ -94,6 +96,19 @@ namespace intrp {
...
+///      A template function that takes as a single argument a const reference
+///      to the \ref DataBoxGroup, and returns a `bool`: `true` if the
+///      element should interpolate in the current state, `false` otherwise.
+///      Often, this will examine the `temporal_id` to determine whether to
+///      interpolate, but it may examine any part of the \ref DataBoxGroup.
+///      If omitted, the behavior is as if the `should_interpolate` function
+///      always returns true. This function is called using the DG element's
+///      \ref DataBoxGroup and is only used if `InterpolateToTarget` is
+///      in the element's action list.
+/// - interpolating_dg_component (required if not `is_sequential`):
```
Comment (markscheel) [inline src/NumericalAlgorithms/Interpolation/InterpolationTarget.hpp]: Maybe "(required if not `is_sequential`)" should be changed to "(used only if not `is_sequential`)"?

File: src/NumericalAlgorithms/Interpolation/InterpolationTarget.hpp [hunk 16]
```diff
@@ -94,6 +96,19 @@ namespace intrp {
...
+///      A type alias for the DG element array that will be interpolating to the
```
Comment (markscheel) [inline src/NumericalAlgorithms/Interpolation/InterpolationTarget.hpp]: Just an observation: if `is_sequential`, then the thing that interpolates to the interpolation target might not be a DG element array.  Not sure if I want a change here, but the current wording might imply that the thing interpolating is always a DG element array, which might be confusing.

File: tests/Unit/NumericalAlgorithms/Interpolation/Test_InterpolateToTarget.cpp [hunk 17]
```diff
@@ -77,23 +77,63 @@ struct initialize_elements_and_queue_simple_actions {
     for (const auto& element_id : element_ids) {
       ActionTesting::next_action<elem_component>(make_not_null(&runner),
                                                  element_id);
+      if (not Metavariables::should_interpolate_value and
+          Metavariables::has_should_interpolate) {
```
Comment (markscheel) [inline tests/Unit/NumericalAlgorithms/Interpolation/Test_InterpolateToTarget.cpp]: Just a question: shouldn't `has_should_interpolate` be checked before `should_interpolate_value`?  It doesn't matter for this test because `should_interpolate_value` is always defined even if `has_should_interpolate` is false, but the ordering struck me as strange.

**Review by markscheel — COMMENTED** [review]
A few minor comments.

File: src/NumericalAlgorithms/Interpolation/Actions/InterpolateToTarget.hpp (hunk 14 above)
Comment (moxcodes) [reply src/NumericalAlgorithms/Interpolation/Actions/InterpolateToTarget.hpp]: I guess I'm not sure whether this should be there anymore -- It's usually used in the should interpolate call. The best thing to do is probably just to note that it invokes the should_interpolate on the box as well.

**Comment by moxcodes** [comment]
Posted the last couple of fixups to documentation (the new ones start with `fixup 2`).
Cheers!

**Comment by kidder** [comment]
I am happy for you to squash when @markscheel gives the go ahead

**Comment by kidder** [comment]
@markscheel please look at

**Review by markscheel — DISMISSED** [review]

**Comment by moxcodes** [comment]
Squashed and rebased on develop.
Cheers!

**Review by nilsdeppe — APPROVED** [review]

**Review by geoffrey4444 — APPROVED** [review]

**Review by markscheel — APPROVED** [review]

**Review by kidder — APPROVED** [review]
